src/GafferScene/PrimitiveVariableQuery.cpp

```diff
@@ -0,0 +1,553 @@
+			);
+		}
+	}
+}
+
+void PrimitiveVariableQuery::hash( const Gaffer::ValuePlug* const output, const Gaffer::Context* const context, IECore::MurmurHash& h ) const
+{
+	Gaffer::ComputeNode::hash( output, context, h );
+
+	if( output == internalObjectPlug() )
+	{
+		const std::string loc = locationPlug()->getValue();
+
+		if( ! loc.empty() )
+		{
+			const ScenePlug* const splug = scenePlug();
+			const ScenePlug::ScenePath path = ScenePlug::stringToPath( loc );
+
+			if( splug->exists( path ) )
+			{
+				h.append( splug->objectHash( path ) );
+			}
+		}
+	}
+	else if( outPlug()->isAncestorOf( output ) )
+	{
+		const Gaffer::ValuePlug* const outputPlug = outPlug( output );
+		const Gaffer::NameValuePlug* const childQueryPlug = queryPlug( outputPlug );
+
+		if(
+			( output == outputPlug->getChild( g_existsPlugIndex        ) ) ||
+			( output == outputPlug->getChild( g_typePlugIndex          ) ) ||
+			( output == outputPlug->getChild( g_interpolationPlugIndex ) ) )
+		{
+			internalObjectPlug()->hash( h );
+			childQueryPlug->namePlug()->hash( h );
+			return;
+		}
+
+		const Gaffer::ValuePlug* const valuePlug = outputPlug->getChild< const Gaffer::ValuePlug >( g_valuePlugIndex );
+		if( valuePlug && ( valuePlug->isAncestorOf( output ) || output == valuePlug ) )
+		{
+			internalObjectPlug()->hash( h );
+			childQueryPlug->namePlug()->hash( h );
+			correspondingPlug(
+				valuePlug,
+				output,
+				static_cast< const Gaffer::ValuePlug* >( childQueryPlug->valuePlug() )
+			)->hash( h );
+		}
+	}
+}
+
+void PrimitiveVariableQuery::compute( Gaffer::ValuePlug* const output, const Gaffer::Context* const context) const
+{
+	if( output == internalObjectPlug() )
+	{
+		// NOTE : cache the primitive at the specified location in the input scene
+		//        as all queries operate on this same primitive.
+
+		IECore::ConstObjectPtr object( IECore::NullObject::defaultNullObject() );
+		const std::string loc = locationPlug()->getValue();
+
+		if( ! loc.empty() )
+		{
+			const ScenePlug* const splug = scenePlug();
+			const ScenePlug::ScenePath path = ScenePlug::stringToPath( loc );
+
+			if( splug->exists( path ) )
+			{
+				object = IECore::runTimeCast< const IECoreScene::Primitive >( splug->object( path ) );
+			}
+		}
+
+		IECore::assertedStaticCast< Gaffer::ObjectPlug >( output )->setValue( object );
+	}
+	else if( outPlug()->isAncestorOf( output ) )
+	{
+		// NOTE : retrieve the cached primitive and query
+
+		const IECore::ConstObjectPtr object = internalObjectPlug()->getValue();
+		assert( object );
+
+		const Gaffer::ValuePlug* const outputPlug = outPlug( output );
+		const Gaffer::NameValuePlug* const childQueryPlug = queryPlug( outputPlug );
+
+		// NOTE : find named primitive variable
+
+		IECoreScene::PrimitiveVariable resultVariable;
+		if( ! object->isEqualTo( IECore::NullObject::defaultNullObject() ) )
+		{
+			const IECoreScene::Primitive* const primitive =
+				IECore::assertedStaticCast< const IECoreScene::Primitive >( object.get() );
```

Comment:
I assume the thinking behind the `assertedStaticCast` is that `__internalObject` has already filtered out non-primitive types, but I'm not sure this really buys us anything. We could simplify by replacing both the NullObject comparison and the assertedStaticCast with a single `if( auto primitive = runTimeCast<const Primitive>( object ) )`, and then also not need any filtering in `__internalObject` (which would remove the nullptr bug).

Reply:
done